server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,905 @@
+		}
+		return {error: e.message};
+	}
+	if (result.result) result = result.result;
+	return {result} || {error: 'Unknown error in database query.'};
+});
+
+
+const ACTIONS = {
+	add: (
+		`REPLACE INTO friends (userid, friend, last_login) VALUES($userid, $friend, $login) ON CONFLICT (userid, friend) ` +
+		`DO UPDATE SET userid = $userid, friend = $friend`
+	),
+	get: (
+		`SELECT *, (SELECT new_name
+			from friend_renames
+			WHERE original_name = friend
+			ORDER BY change_date desc
+			LIMIT 1) new_name
+		FROM friends WHERE userid = ? LIMIT ?`
+	),
+	// may look duplicated, but you pass in [userid1, userid2, userid2, userid1]
+	delete: `DELETE FROM friends WHERE (userid = ? AND friend = ?) OR (userid = ? AND friend = ?)`,
+	getSent: `SELECT receiver, sender FROM friend_requests WHERE sender = ?`,
+	getReceived: `SELECT receiver, sender FROM friend_requests WHERE receiver = ?`,
+	insertRequest: `INSERT INTO friend_requests(sender, receiver, sent_at) VALUES(?, ?, ?)`,
+	deleteRequest: `DELETE FROM friend_requests WHERE sender = ? AND receiver = ?`,
+	findFriendship: `SELECT * FROM friends WHERE (friend = $user1 AND userid = $user2) OR (userid = $user1 AND friend = $user2)`,
+	findRequest: `SELECT * FROM friend_requests WHERE (sender = $user1 AND receiver = $user2) OR (sender = $user2 AND receiver = $user1)`,
+	renameFriend: `UPDATE OR IGNORE friends SET friend = $newID WHERE friend = $oldID`,
+	renameUserid: `UPDATE OR IGNORE friends SET userid = $newID WHERE userid = $oldID`,
+	rename: `REPLACE INTO friend_renames (original_name, new_name, change_date) VALUES(?, ?, ?)`,
+	login: `UPDATE friends SET last_login = ? WHERE friend = ?`,
+	checkLastLogin: `SELECT last_login FROM friends WHERE friend = ?`,
+	deleteLogin: `UPDATE friends SET last_login = null WHERE friend = ?`,
+	expire: (
+		`DELETE FROM friend_requests WHERE EXISTS` +
+		`(SELECT sent_at FROM friend_requests WHERE should_expire(sent_at) = 1)`
+	),
+	hidingLogin: `SELECT * FROM friend_settings WHERE name = ? AND send_login_data = 1`,
+	hideLogin: `REPLACE INTO friend_settings (name, send_login_data) VALUES (?, 1)`,
+	showLogin: `DELETE FROM friend_settings WHERE name = ? AND send_login_data = 1`,
+};
+
+
+const TRANSACTIONS: {[k: string]: (input: any[]) => DatabaseResult} = {
+	rename: requests => {
+		for (const request of requests) {
+			const [oldID, newID] = request;
+			const oldFriends = statements.get.all(oldID, MAX_FRIENDS).map(item => item.friend);
+			const newRequests = statements.getReceived.all(oldID);
+			const sentRequests = statements.getSent.all(oldID);
+			if (newRequests.length || sentRequests.length) {
+				throw new Chat.ErrorMessage(`You have pending friend requests, and so cannot transfer friends.`);
+			}
+			if (!oldFriends.length) {
+				throw new Chat.ErrorMessage(`You have no friends to transfer.`);
+			}
+			if (oldFriends.includes(newID)) {
+				// unfriend the two
+				statements.delete.run({userid: newID});
+			}
```

Comment:
There is no need to fetch all these rows. You can do `SELECT count(*)` for `newRequests`/`sentRequests`, but it is unclear to me why you can't transfer with stuff pending? For `oldFriends`, just do the delete unconditionally